### PR TITLE
bpo-45924: Fix asyncio incorrect traceback when future's exception is raised multiple times

### PR DESCRIPTION
**Pull request by kumaraditya303**
https://bugs.python.org/issue45924

### CODE REVIEW

**Comment by kumaraditya303** [comment]
With this patch traceback:

```console
Traceback (most recent call last):
  File "D:\cpython\main.py", line 17, in main
    print(await fut)
          ^^^^^^^^^
TypeError: 42
Traceback (most recent call last):
  File "D:\cpython\main.py", line 17, in main
    print(await fut)
          ^^^^^^^^^
TypeError: 42
Traceback (most recent call last):
  File "D:\cpython\main.py", line 17, in main
    print(await fut)
          ^^^^^^^^^
TypeError: 42

```

File: Modules/_asynciomodule.c [hunk 1]
```diff
@@ -646,6 +650,16 @@ future_get_result(FutureObj *fut, PyObject **result)
 
     fut->fut_log_tb = 0;
     if (fut->fut_exception != NULL) {
+        if (fut->fut_exception_tb != NULL) {
```
Comment (kumaraditya303) [inline Modules/_asynciomodule.c]: Hi @erlend-aasland Would you like to review this? Thanks (there may be some ref leaks)

Comment (erlend-aasland) [reply Modules/_asynciomodule.c]: Hi, Kumar! I'll have a look hopefully in a couple of days.

File: Modules/_asynciomodule.c [hunk 2]
```diff
@@ -646,6 +650,16 @@ future_get_result(FutureObj *fut, PyObject **result)
...
+            if (PyException_SetTraceback(fut->fut_exception, fut->fut_exception_tb) < 0)
+            {
+                Py_DECREF(fut->fut_exception_tb);
```
Comment (erlend-aasland) [inline Modules/_asynciomodule.c]: Note: `PyException_SetTraceback` takes ownership of the object, so we have no decref responsibility on our side. Also, `fut->fut_exception_tb` points to some object that you just decref'ed. Probably not what you want :)

Comment (kumaraditya303) [reply Modules/_asynciomodule.c]: The strange thing is that on windows tests are passing but failing on Linux

**Comment by erlend-aasland** [comment]
BTW, GitHub does not play well with force pushes (contrary to GitLab). If you use `git pull --no-ff main` instead of `git rebase main`, changes will play well with the GitHub Pull Request ui, and the review process will be a lot easier. If you force push, review comments/discussions may end up in limbo.

**Comment by kumaraditya303** [comment]
Marking this as ready for review as Windows is passing but I don't know what is meant by this 
```console
1 test altered the execution environment:
    test_asyncio
```

File: Lib/test/test_asyncio/test_futures2.py [hunk 3]
```diff
@@ -1,9 +1,32 @@
...
 
 class FutureTests(unittest.IsolatedAsyncioTestCase):
+    async def test_future_traceback(self): 
+        async def raise_exc():
+            raise TypeError(42)
+        future = asyncio.create_task(raise_exc())
+        for _ in range(10):
+            try:
+                await future
+            except TypeError:
```
Comment (sobolevn) [inline Lib/test/test_asyncio/test_futures2.py]: Please, add `else: self.fail('TypeError not raised')`

File: Lib/test/test_asyncio/test_futures2.py [hunk 4]
```diff
@@ -1,9 +1,32 @@
...
+    await future
+    ^^^^^^^^^^^^
+  File "{__file__}", line 9, in raise_exc
+    raise TypeError(42)
+    ^^^^^^^^^^^^^^^^^^^
+TypeError: 42
+"""
+        )
+            self.assertEqual(tb, expected)
+
```
Comment (asvetlov) [inline Lib/test/test_asyncio/test_futures2.py]:
```suggestion
            else: 
                self.fail('TypeError not raised')

```

File: Lib/test/test_asyncio/test_futures2.py (hunk 3 above)
Comment (asvetlov) [reply Lib/test/test_asyncio/test_futures2.py]: Done

**Comment by asvetlov** [comment]
Fixed.
The test is very weak though:
1. Adding/removing source lines before the test breaks it.
2. Tested tracebacks are the same. I prefer that the test reflects *new* traceback substitution.
3. Textual representation of stack trace was changed recently, new improvements can break the test again.
Perhaps tracebacks should be analyzed, no printed exceptions.

**Comment by asvetlov** [comment]
New test is a little harder to read but much stablier

**Comment by asvetlov** [comment]
Interesting. `test_tasks` alters the environment by this PR.
The main branch has no such defect.
Should double check `_asynciomodule.c` changes, I have no other idea.

I use `./python -We -m test.regrtest -vvv test_asyncio -m test_tasks` to check

File: Lib/test/test_asyncio/test_futures2.py [hunk 5]
```diff
@@ -1,13 +1,68 @@
...
+                if found != -1:
+                    marker = line[found + len(mark):].strip()
+                    if marker == '"':
+                        # skip 'mark' variable definition from above
+                        continue
+                    markers[marker] = offset + pos
+
+        init_markers()
+
+        def analyze_traceback(tb, await_marker):
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py]: The traceback tests do this by capturing the line number in a variable and using that. See

https://github.com/python/cpython/blob/e63894b3eed8ad2dd7690695f7f07bfbff59c05a/Lib/test/test_traceback.py#L386

and

https://github.com/python/cpython/blob/e63894b3eed8ad2dd7690695f7f07bfbff59c05a/Lib/test/test_traceback.py#L395

**Comment by iritkatriel** [comment]
> Interesting. `test_tasks` alters the environment by this PR. The main branch has no such defect. Should double check `_asynciomodule.c` changes, I have no other idea.
> 
> I use `./python -We -m test.regrtest -vvv test_asyncio -m test_tasks` to check

It's due to the changes in RunCoroutineThreadsafeTests. When I revert them this problem goes away.

File: Lib/test/test_asyncio/test_tasks.py [hunk 6]
```diff
@@ -3288,13 +3288,12 @@ async def coro(s):
             self.one_loop.call_exception_handler.assert_not_called()
 
 
-class RunCoroutineThreadsafeTests(test_utils.TestCase):
+class RunCoroutineThreadsafeTests(unittest.IsolatedAsyncioTestCase):
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_tasks.py]: These changes need to be reverted - they are unrelated to the current issue, and they also caused a test failure.

File: Modules/_asynciomodule.c [hunk 7]
```diff
@@ -655,6 +659,15 @@ future_get_result(FutureObj *fut, PyObject **result)
...
     if (fut->fut_exception != NULL) {
+        if (fut->fut_exception_tb != NULL) {
+            if (PyException_SetTraceback(fut->fut_exception, fut->fut_exception_tb) < 0)
+            {
+                return -1;
+            }
+        }
+        else {
+            PyException_SetTraceback(fut->fut_exception, Py_None);
+        }
```
Comment (iritkatriel) [inline Modules/_asynciomodule.c]:
```suggestion
        PyObject *tb = fut->fut_exception_tb;
        if (tb == NULL) {
            tb = Py_None;
        }
        if (PyException_SetTraceback(fut->fut_exception, tb) < 0) {
            return -1;
        }
```

File: Lib/test/test_asyncio/test_futures2.py [hunk 8]
```diff
@@ -1,13 +1,68 @@
...
 import unittest
+import traceback
 
 
 def tearDownModule():
     asyncio.set_event_loop_policy(None)
 
 
 class FutureTests(unittest.IsolatedAsyncioTestCase):
+    maxDiff = None
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py]: Why do we need this?

File: Lib/test/test_asyncio/test_futures2.py [hunk 9]
```diff
@@ -1,13 +1,68 @@
...
+
+    async def test_future_traceback(self):
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py:14]: We can have a simple test that awaits in a loop and checks that the traceback is always the same (the example from the BPO, essentially).

We can also have this test, where each await is from a different place, but as I mentioned in https://github.com/python/cpython/pull/30274/files#r830857535 it doesn't need all this complexity.

**Comment by iritkatriel** [comment]
> > Interesting. `test_tasks` alters the environment by this PR. The main branch has no such defect. Should double check `_asynciomodule.c` changes, I have no other idea.
> > I use `./python -We -m test.regrtest -vvv test_asyncio -m test_tasks` to check
> 
> It's due to the changes in RunCoroutineThreadsafeTests. When I revert them this problem goes away.

The tests are failing on the buildbots (but not locally for me) after reverting this change, so I don't know if that was it.

**Comment by gvanrossum** [comment]
Looks like we need to understand better what the failure means. "test_asyncio changed the environment" ??? Asking @vstinner

**Comment by vstinner** [comment]
> Looks like we need to understand better what the failure means. "test_asyncio changed the environment" ??? Asking @vstinner

The message can mean anything. The logs should tell you what changed. It can be an unhandled exception, a file which is created but not removed, a process which is created but doesn't complete, etc.

Usually, I try to reproduce the issue locally and run tests in verbose mode.

Also, ``./python -m test.bisect_cmd -o bisect --fail-env-changed test_asyncio [...]`` helps to identify which test methods changes the environment. At the end, you should get a single test method name in the ``bisect`` file.

**Comment by gvanrossum** [comment]
> The message can mean anything.

What I meant was, what is the "execution environment" that is being altered here (I presume it's not just the set of UNIX environment variables :-) and how is it tested for changes. I guess I'll have to read the [source](https://github.com/python/cpython/blob/main/Lib/test/libregrtest/main.py#L392)...

But a grep reminds me of another asyncio test that has an explicit comment
```py
# To prevent a warning "test altered the execution environment"
def tearDownModule():
    asyncio.set_event_loop_policy(None)
```
So maybe there's another test that now requires something like that. (@kumaraditya303 any ideas?)

**Comment by vstinner** [comment]
Result of test.bisect_cmd:

```
$ ./python -m test --fail-env-changed test_asyncio -v -m test.test_asyncio.test_tasks.RunCoroutineThreadsafeTests.test_run_coroutine_threadsafe_task_factory_exception
== CPython 3.11.0a7+ (heads/pr/30274:9c9f0e68d7, May 3 2022, 22:15:56) [GCC 11.3.1 20220421 (Red Hat 11.3.1-2)]
== Linux-5.17.4-200.fc35.x86_64-x86_64-with-glibc2.34 little-endian
== cwd: /home/vstinner/python/main/build/test_python_747606æ
== CPU count: 8
== encodings: locale=UTF-8, FS=utf-8
0:00:00 load avg: 1.81 Run tests sequentially
0:00:00 load avg: 1.81 [1/1] test_asyncio
test_run_coroutine_threadsafe_task_factory_exception (test.test_asyncio.test_tasks.RunCoroutineThreadsafeTests.test_run_coroutine_threadsafe_task_factory_exception)
Test coroutine submission from a thread to an event loop ... 

Warning -- threading_cleanup() failed to cleanup 0 threads (count: 0, dangling: 2)
Warning -- Dangling thread: <Thread(asyncio_0, stopped 140244085950016)>
Warning -- Dangling thread: <_MainThread(MainThread, started 140244334782272)>

ok

----------------------------------------------------------------------
Ran 1 test in 0.073s

OK
test_asyncio failed (env changed)

== Tests result: ENV CHANGED ==

1 test altered the execution environment:
    test_asyncio

Total duration: 802 ms
Tests result: ENV CHANGED
```

**Comment by gvanrossum** [comment]
Thanks -- that error occurs even if I remove the changes to futures.py, so it must be something in the changes to `_asynciomodule.c`. @kumaraditya303 Maybe you can look into that?

**Comment by gvanrossum** [comment]
@kumaraditya303, Victor said in Discord "test_asyncio should wait until its threads complete 🙂 otherwise, it has nasty side effects on following tests making them less reliable"

File: Lib/test/test_asyncio/test_futures2.py (hunk 5 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py]: It is much simplified now.

File: Lib/test/test_asyncio/test_futures2.py (hunk 8 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py]: Removed

File: Lib/test/test_asyncio/test_futures2.py (hunk 9 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py:14]: Test is simplified now.

**Comment by kumaraditya303** [comment]
> Maybe you can look into that?

Well, I fixed it by not keeping a reference to the traceback which presumably keeps a frame and that frame keeps a thread state alive hence the altered environment error, but now it's fixed.

File: Lib/test/test_asyncio/test_futures2.py [hunk 10]
```diff
@@ -1,13 +1,31 @@
...
 class FutureTests(unittest.IsolatedAsyncioTestCase):
+
+    async def test_future_traceback(self):
+
+        async def raise_exc():
+            raise TypeError(42)
+
+        future = asyncio.create_task(raise_exc())
+
+        for _ in range(5):
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py:21]:
```suggestion
        for i in range(5):
```

File: Lib/test/test_asyncio/test_futures2.py [hunk 11]
```diff
@@ -1,13 +1,31 @@
...
+            try:
+                await future
+            except TypeError as e:
+                tb = ''.join(traceback.format_tb(e.__traceback__))
+                self.assertEqual(tb.count("await fut"), 1)
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py]:
```suggestion
                if i > 0:
                    self.assertEqual(tb, prev_tb)
                prev_tb = tb
```

**Review by iritkatriel — CHANGES_REQUESTED** [review]
Cool!

You could just compare the whole traceback, something like this.

File: Lib/test/test_asyncio/test_futures2.py (hunk 11 above)
Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py]: The existing test already seems sufficient and it makes it explicit that "await fut" should only appear once in traceback but thanks for suggestion.

**Comment by kumaraditya303** [comment]
> Sorry, I didn't mean to press "request changes", it was just a suggestion.

No worries, but I think the current test is already sufficient and the test is explicit about the occurrence of "await foo" for which the issue is all about.

**Comment by iritkatriel** [comment]
> > Sorry, I didn't mean to press "request changes", it was just a suggestion.
> 
> No worries, but I think the current test is already sufficient and the test is explicit about the occurrence of "await foo" for which the issue is all about.

My unease with this test is that the "await fut" string comes from the library rather than the test. If the code in the library changes the code could break (or not) in a way that is hard to predict.

**Comment by kumaraditya303** [comment]
I think the "await fut" comes from the test itself and has nothing to do with the library. Could you provide a example where this might break ?

**Comment by iritkatriel** [comment]
> I think the "await fut" comes from the test itself and has nothing to do with the library. Could you provide a example where this might break ?

IIUC "await fut" is the source line from the asyncio library. I don't see it in the test.

**Comment by iritkatriel** [comment]
> Could you provide a example where this might break ?

For instance, if someone renames the variable "fut" in the library, the test should not break but it will because it's asserting on implementation details.

**Review by iritkatriel — APPROVED** [review]
Ok, that’s clearer.

I still would have written a test that checks the whole traceback is unchanged (because it should be) but up to you.

**Review by gvanrossum — APPROVED** [review]
What's up with that one failing test? Fluke? I'll see if I can re-run it.

**Comment by gvanrossum** [comment]
Re-running that job didn't help. I don't know what it does. github-token not supplied -- what's that?

**Comment by arhadthedev** [comment]
> I don't know what it does. github-token not supplied -- what's that?

https://github.com/python/cpython/blob/14fea6b4d25658bc00adbb97dd40ea3d3e6843ad/.github/workflows/project-updater.yml#L17

It seems that `ADD_TO_PROJECT_PAT` secret variable is empty.

File: Lib/test/test_asyncio/test_futures2.py [hunk 12]
```diff
@@ -1,13 +1,31 @@
...
 import unittest
+import traceback
 
 
 def tearDownModule():
     asyncio.set_event_loop_policy(None)
 
 
 class FutureTests(unittest.IsolatedAsyncioTestCase):
+
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py:13]: I'm assuming this test runs both the C and the python versions?

Comment (gvanrossum) [reply Lib/test/test_asyncio/test_futures2.py:13]: I think there's a separate run of the tests that disables `_asyncio`? I have no idea where this is done though.

Comment (iritkatriel) [reply Lib/test/test_asyncio/test_futures2.py:13]: If I revert the C fix the test fails. If I revert the python fix it doesn't.

Comment (gvanrossum) [reply Lib/test/test_asyncio/test_futures2.py:13]: Yeah, you have to use a pattern that creates two separate test classes, similar to [this](https://github.com/python/cpython/blob/14fea6b4d25658bc00adbb97dd40ea3d3e6843ad/Lib/test/test_asyncio/test_futures.py#L638-L667) (but you don't need to test with a Python subclass of `CFuture`).

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py:13]: I had thought that all the tests were run with C accelerator and pure Python but seems that's not the case. I'll see to it.

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py:13]: Fixed in https://github.com/python/cpython/pull/30274/commits/9d6bae0b281e764a8f83f2e91b9dfff37588917c

File: Lib/test/test_asyncio/test_futures2.py [hunk 13]
```diff
@@ -1,13 +1,32 @@
...
+        for _ in range(5):
+            try:
+                await future
+            except TypeError as e:
+                tb = ''.join(traceback.format_tb(e.__traceback__))
+                self.assertEqual(tb.count("await future"), 1)
+            else:
+                self.fail('TypeError was not raised')
+
     async def test_recursive_repr_for_pending_tasks(self):
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py:40]: Now this test is running twice, but it doesn't use ``cls`` so the two executions are identical?

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py:40]: `test_recursive_repr_for_pending_tasks` does not uses it so two runs are identical.

Comment (iritkatriel) [reply Lib/test/test_asyncio/test_futures2.py:40]: So it should be in a different class so it runs once.

Comment (kumaraditya303) [reply Lib/test/test_asyncio/test_futures2.py:40]: done.

File: Lib/test/test_asyncio/test_futures2.py [hunk 14]
```diff
@@ -1,13 +1,34 @@
...
+
+        for _ in range(5):
+            try:
+                await future
+            except TypeError as e:
+                tb = ''.join(traceback.format_tb(e.__traceback__))
+                self.assertEqual(tb.count("await future"), 1)
+            else:
+                self.fail('TypeError was not raised')
+
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py:29]:
```suggestion

@unittest.skipUnless(hasattr(tasks, '_CTask'),
                       'requires the C _asyncio module')
  class CFutureTests(FutureTests, unittest.IsolatedAsyncioTestCase):
      cls = tasks._CTask
 
  class PyFutureTests(FutureTests, unittest.IsolatedAsyncioTestCase):
      cls = tasks._PyTask
```

File: Lib/test/test_asyncio/test_futures2.py [hunk 15]
```diff
@@ -21,6 +42,13 @@ async def func():
         # exact comparison for the whole string is even weaker.
         self.assertIn('...', repr(await asyncio.wait_for(func(), timeout=10)))
 
+@unittest.skipUnless(hasattr(tasks, '_CTask'),
+                     'requires the C _asyncio module')
+class CFutureTests(FutureTests, unittest.IsolatedAsyncioTestCase):
+    cls = tasks._CTask
+
+class PyFutureTests(FutureTests, unittest.IsolatedAsyncioTestCase):
+    cls = tasks._PyTask
```
Comment (iritkatriel) [inline Lib/test/test_asyncio/test_futures2.py]:
```suggestion

```

**Review by iritkatriel — APPROVED** [review]
I'd put the subclasses right after the base class, like so.

**Pull request merged**